### ISSUE TITLE
Strong 10b KataGo net

### ISSUE THREAD

**Issue by Friday9i** [lightvector/KataGo#88]
Hello
I'm doing an independant selfplay + training run of KataGo nets, simultaneously doing selfplay with a 10b128f and a 30b256f net, and training them in parallel ; -). Recently, I added a 15b192f (still weaker than the 10b net but coming close).
I began from random nets and was doing selfplay with extremely low visits: only 5 cheap and 30 normal visits! And it worked well until the 30b net reached a level approximately in the middle between the official g104 10b and 15b nets (ie probably around ~LZ120's strength). Then it stagnated for some time, so I upped the visits progressively to 30 cheap / 120 normal visits. And as imrpovements are slowing down, I'm upping visits to more classic 100/600 or more visits...
30b was improving again but 10b stagnated until I decreased the LR to 0.3 (it gained ~100 Elo very quickly). So I'm currently reducing the LR of 10b progressively while I keep the 30b LR at 1 as it still improves well.
Today, after 1.5 months in the run (on 2 GPUs: RTX2070+RTX2080), here are the results:

**Trained 10b net:**
It is ~200 Elo stronger than the official g104 10b net on all board sizes from 6x6 to 19x19.
Here are the detailed results running gatekeeper with random odd komi (within ~1 point from fair komi) and up to 200 games for each board sizes, with 100 visits:
6x6 | 7x7 | 8x8 | 9x9 | 11x11 | 13x13 | 19x19
69% | 81% | 89% | 72% | 71% | 75% | 80%
With more visits, results would probably be even more pronounced in favor of this net (but I didn't test it).
This net is also better than g104 15b (and much faster of  course!) on board sizes from 6x6 to ~10x10, and probably still competitive up to 19x19 on time parity. Here are the results with 100 visits for both nets (on same board sizes):
67% | 78% | 75% | 62% | 41% | 39% | 20%
Here is the 10b net (I'll publish new versions in the coming weeks if they get significantly stronger):
[model.txt.gz](https://github.com/lightvector/KataGo/files/3929223/model.txt.gz)

**Trained 15b net:** it is very recent and still a bit weaker than 10b (hence little interest for the moment) 
**Trained 30b net:** it is clearly stronger than the 10b, especially on boards bigger than 13x13, but still "only" around g104b 15b strength (ie close to LZ145/LZ150) while much slower. _BTW, it's just amazing to generate from scratch a net around LZ145's strength on a single computer in 1.5 months!!!_
I'm now switching to a mix training and selfplay to see if it accelerates the training (using 10b for deep selfplay games mainly on small board sizes while I generate 100/600 visits 30b games mainly for large boards sizes): we'll see if it is efficient!

**Comment by cryptpark** [comment]
@Friday9i thanks! very interesting! 30b256f is not a typo? 30b or 20b?

**Comment by Friday9i** [comment]
Not a typo, it's a 30b256f network indeed ; -)
20b256f already exists, so I thought it would be interesting to test a slightly larger net and see if it can go higher. And it's still manageable in terms of size and speed: it takes around 2 hours to train it, and it plays reasonably fast in selfplay, so an attractive size I think.

**Comment by cryptpark** [comment]
@Friday9i yes it is interesting, thanks!

**Comment by kaorahi** [comment]
Great for CPU-only users! Can I use it as a build-in network of my [GUI](https://github.com/kaorahi/lizgoban/)?

If possible, would you provide saved_model/{model.config.json,variables} like the official g104? They are required for libtensorflow backend (https://github.com/lightvector/KataGo/issues/14#issuecomment-549142392).

**Comment by Friday9i** [comment]
Unfortunately I had a crash and went back to a backup 24 hours old, and I lost this folder...
Good news is I have a more recent 10b which seems close in strength to the previous one (maybe slightly weaker, but within noise, so comparable). Here is the full zipped folder (too big, so here is a link):
https://filetransfer.io/data-package/YiGnULsu

**Comment by kaorahi** [comment]
Thanks! I got converted *.pb from Dec8 version successfully for libtensorflow backend.

I cannot combine Dec8 version with model.txt.gz in Dec5 version. The converter seems to depend only on `saved_model/{variables*,model.config.json}`. Anyway, I will try Dec8 version as my default network for the game analysis for a while.

Would you permit to use / distribute your networks as built-in ones inside a packed *.exe? I will refer to the origin of them in the release notes like [this](https://github.com/kaorahi/lizgoban/releases/tag/v0.2.0).

**Comment by Friday9i** [comment]
Yes no problem to use and redistribute the nets I trained!
(if lightvector is ok also, as it is done with KataGo of course)

**Comment by lightvector** [comment]
Seems okay to me. It's all open-source anyways. :)

**Comment by kaorahi** [comment]
Then I will use Dec5 version in the next release. I enjoy watching its wins against ELFv1 in time parity games on my machine. Thanks.

**Comment by l1t1** [comment]
useful work

**Comment by l1t1** [comment]
i am running a test
```
D:\tool\go_gui\gogui-twogtp -white "D:\KataGo-1.1\katago12_cuda gtp -model d:\model1205.txt -config D:\gtp_example.cfg" -black "D:\KataGo-1.1\katago12_cuda gtp -model d:\model1209.txt -config D:\gtp_example.cfg"  -games 20 -sgffile katago12cuda_120509 -auto -komi 7.5
```

**Comment by Friday9i** [comment]
@l1t1 Gatekeeper is way faster than Gogui for KataGo : -).
You can use the following text in a bat file (on windows):
"C:\katago gatekeeper -rejected-models-dir D:/Games/RejectedNets -accepted-models-dir D:/Games/AcceptedNets -sgf-output-dir D:/Games/SGFs -test-models-dir D:/Games/TestedNets -config-file GATEKEEPER.cfg >> Result.txt"

**Comment by l1t1** [comment]
@Friday9i thanks

**Comment by kaorahi** [comment]
very aggressive H4 game by Dec5 network (W, 30sec) against leela 0.11.0 (B, 10sec) with modified katago (https://github.com/kaorahi/KataGo/tree/keep_uneasy)

<img src="https://user-images.githubusercontent.com/38910552/70800027-8809ff80-1dee-11ea-9b4e-474f6d3ad8db.png" width=49%> <img src="https://user-images.githubusercontent.com/38910552/70800422-93115f80-1def-11ea-966a-8cbaac25eae6.png" width=49%>
[L0110_H4.sgf.txt](https://github.com/lightvector/KataGo/files/3960716/L0110_H4.sgf.txt)

**Comment by Friday9i** [comment]
**2019-12-20:** here is a link to a new 10b apparently significantly stronger:
https://filetransfer.io/data-package/kxPmeRyR
And the model itself (without additional folders) is here: 
[model.txt.gz](https://github.com/lightvector/KataGo/files/3986000/model.txt.gz)

**Comment by kaorahi** [comment]
Thanks again :) I use your network by default for game analyses now and I will pack it into the next release of my GUI. It is good at large capturing races for a 10b network in my impression.

**Comment by Friday9i** [comment]
Nice!
If you run some benchmarks / matches against other nets (LZ, Elf, other KataGo nets ...), would you mind to share the results here? Thanks a lot

**Comment by kaorahi** [comment]
My PC is too slow for benchmarks... I have only a portable laptop without GPU. (So I appreciate your small & strong networks.)

**Comment by cryptpark** [comment]
gt 610 is also not a strong card, but maybe it's interesting:
50 games KataGo 10x128-s1108 - KataGo 10x128-s4073 (gt 610, 5 sec/move), new network won 42:8 (21 b and 21 w)
50 games KataGo 15x192-s2973 - KataGo 10x128-s4073 (gt 610, 5 sec/move), new network won 32:18 (15 b and 17 w)
50 games KataGo 20x192-s4479 - KataGo 10x128-s4073 (gt 610, 5 sec/move), new network lost 8:42 (1 b and 7 w)

**Comment by Friday9i** [comment]
Thanks a lot @cryptpark 
Are those games only on 19x19 with 7.5 komi, or spread on different board sizes? How many visits approximately with 5s/move?
Anyway, it seems to confirm this first 10b I published (s4073) is well above the official g104 10b net and also slightly above the official g104 15b net ; -).
And the last net published yesterday (s4588) should be even better, hopefully closer to g104 20b on 19x19 (and better on boards below ~13x13).

**Comment by cryptpark** [comment]
@Friday9i Yes, on 19x19 with 7.5 komi. first 20 moves Root visits (gtp.log) - min 147 to max 789 (10x128-s4073). in an hour I will start the match s4588 - 20x256-s4479. Thanks again for your interesting work! and good luck !!!

**Comment by cryptpark** [comment]
It is fantastic! new network on the same level with 20x256! now 19:23

**Comment by Friday9i** [comment]
Almost too good to be true! Let's wait for some more games, but that's a good start ;-)

**Comment by cryptpark** [comment]
50 games KataGo 10x128-s4588 - KataGo 20x256-s4479 (gt 610, 5 sec/move), new network lost 22:28 (8 b and 14 w)

**Comment by l1t1** [comment]
suggest release it as a new offical weight, so that we can have a new benchmark just like
http://www.yss-aya.com/cgos/19x19/cross/kata12_104b20_v800.html

**Comment by cryptpark** [comment]
@Friday9i I want to make a match against SAI 9x9. What parameters to use for maximum KataGo power?

**Comment by Friday9i** [comment]
With a good GPU, you should use 10 to 20 threads, but for a weak GPU, 3 or 5 threads may be better.
And equivalent time I would say

**Comment by cryptpark** [comment]
and -r, -b and other options?

**Comment by Friday9i** [comment]
The -r option is managed by the config.gtp file, and I don't know -b : -)

**Comment by cryptpark** [comment]
it's from LZ, I thought the same thing in KataGo. Thank you!

**Comment by cryptpark** [comment]
100 games 9x9, KataGo 10x128-s1108 - SAI (gt 610, 5 sec/move) 30:70 (KataGo b 9:41 and KataGo w 21:29)
100 games 9x9, KataGo 10x128-s4588 - SAI (gt 610, 5 sec/move) 65:35 (KataGo b 23:27 and KataGo w 42:8)

**Comment by Friday9i** [comment]
Eh eh, cool ; -). Which version of SAI?

**Comment by cryptpark** [comment]
the latest release of SAI, 0.17.4 https://github.com/sai-dev/sai/releases/

**Comment by 22nsuk** [comment]
@Friday9i How's the progress going on for 30 blocks? I want you to release this as well.
And I'd appreciate it if you could summarize and explain how to train katago. Is that possible?
Thank you for sharing a good model.

**Comment by cryptpark** [comment]
200 more games 9x9, sai with additional parameters:
100 games KataGo 10x128-s1108 - SAI (--lambda 0 --mu 0 --nrsymm), (gt 610, 5 sec/move) 55:45 (KataGo b 17:33 and KataGo w 38:12)
100 games KataGo 10x128-s4588 - SAI (--lambda 0 --mu 0 --nrsymm --puct 0), (gt 610, 5 sec/move) 84:16 (KataGo b 39:11 and KataGo w 45:5)

**Comment by cryptpark** [comment]
new release SAI 0.17.5
100 games KataGo 10x128-s4588 - SAI (--lambda 0 --mu 0 --nrsymm), (gt 610, 5 sec/move) 48:52 (KataGo b 10:40 and KataGo w 38:12)

**Comment by cryptpark** [comment]
100 games  10 sec/move:
KataGo 10x128-s4588 - SAI (--lambda 0 --mu 0 --nrsymm), (gt 610) 51:49 (KataGo b 11:39 and KataGo w 40:10)

**Comment by cryptpark** [comment]
50 games 5 sec/move:
KataGo 10x128-s4588 - ELF v2 (gt 610) 37:13 (KataGo b 16:9 and KataGo w 21:4)

**Comment by Friday9i** [comment]
Here is a new version of the 10b, around 20 Elo stronger on average than v2. Here are the result from 200 games per board size, vs g104 10b, 15b and 20b:

10b v3 | 6x6 | 7x7 | 8x8 | 9x9 | 11x11 | 13x13 | 19x19
-- | -- | -- | -- | -- | -- | -- | --
vs 10b g104 | 66% | 93% | 85% | 76% | 71% | 81% | 85%
vs 15b g104 | 72% | 74% | 74% | 65% | 53% | 40% | 26%
vs 20b g104 | 67% | 67% | 66% | 44% | 19% | 7% | 2%

Link to the net and the full folder are there: https://filetransfer.io/data-package/SQA41FYS

**Comment by luckyangman** [comment]
Excellent work. Not tested with g161?

**Comment by kennychenfs** [comment]
Where is g161?

**Comment by luckyangman** [comment]
https://drive.google.com/file/d/1caHLk2cuDX9EeZxluk_4nUy9OAvpwEhD/view

**Comment by luckyangman** [comment]
https://drive.google.com/file/d/1Xa3mpB3Wjen67-nYxETMygeLFrqokPWp/view

**Comment by kennychenfs** [comment]
Is g161 the third run?
Where can I find something like "https://d3dndmfyhecmj0.cloudfront.net/" ?

**Comment by Friday9i** [comment]
And what is g161...?

**Comment by luckyangman** [comment]
g161-b15c192-s360540160-d118721852.txt.gz

**Comment by Friday9i** [comment]
Yeah ok it's a net, but what is this net, how is it done, what are the differences, who did it,...?

**Comment by luckyangman** [comment]
lightvector

**Comment by luckyangman** [comment]
In the discord on December 7, 2019

**Comment by kaorahi** [comment]
Do we need some unpublished version of KataGo to use this network?
Is it related to this "handicap game style"??
https://lifein19x19.com/viewtopic.php?f=18&t=17183

~~~
$ ./katago gtp -model g161-b15c192-s360540160-d118721852.txt.gz -config gtp_my.cfg
KataGo v1.2
terminate called after throwing an instance of 'StringError'
  what():  Error loading or parsing model file /XXX/g161-b15c192-s360540160-d118721852.txt.gz: b15c192-s360540160-d118721852: model found unsupported version 7
~~~

**Comment by lightvector** [comment]
Yes you do, and no it's not related to handicap style, it was intended for one specific test that an LZ discord user wanted to do regarding opening move variety in even games, and nothing else.
Please don't bother actually using this network - I don't know why why anyone would be interested in it for anything else. It's unofficial, it's not the strongest, it doesn't specially support any interesting new features, etc.

**Comment by luckyangman** [comment]
G161 Although this weight(network ) is not strong,  the handicap  be very good. No 33 is amazing.Like the strongest 10b. Let's be interested in testing.

**Comment by cryptpark** [comment]
No 33 - what is it?

**Comment by cryptpark** [comment]
50 games KataGo 10x128-s5350 - KataGo 20x256-s4479 (gt 610, 5 sec/move), new network lost 13:37 (6 b and 7 w). s4588 - 22:28! curiously...

**Comment by luckyangman** [comment]
> No 33 - what is it?

Checkerboard position. Near star positions at 4 corners

**Comment by cryptpark** [comment]
50 games 19x19 KataGo 10x128-s5350 - KataGo 15x192-s2973 (gt 610, 5 sec/move), new network won 36:14 (19 b and 17 w)

**Comment by cryptpark** [comment]
50 games 10x128 g170e - 10x128 s4588 (19x19, KataGo-v1.3, gt 610, 5 sec/move), new network won 38:12 (20 b and 18 w)

**Comment by Friday9i** [comment]
Yeah thanks @cryptpark, the new g170 10b is indeed stronger than mine on 19x19.
But mine seems better on small boards: if you have a chance to test on 9x9 that would be interesting, and maybe 7x7, and also 11x11 and 13x13.

**Comment by cryptpark** [comment]
ok, same networks? or s5350

**Comment by Friday9i** [comment]
As you prefer (both would be interesting if you can)

**Comment by cryptpark** [comment]
it will take several days

**Comment by Friday9i** [comment]
If you are motivated, perfect!
Otherwise, I can do it much faster with 500 visits/move (a few hours I guess), tell me

**Comment by cryptpark** [comment]
I put it at night, if it makes sense in your opinion, I will. but I'm more interested in big boards. sorry, I can not compile kato without a 19x19 restriction. if anyone did this i would be just happy !!!

**Comment by Friday9i** [comment]
Cool if you can test indeed. I'm interested in big boards too, I'd like to see KataGo trained a bit for large boards up to 37x37 ; -). And if someone can compile it for Windows and boards larger than 19x19, I'm interested also!

**Comment by cryptpark** [comment]
I just don’t know how to compile. perhaps in one or two source files you need to change the value 19x19. if anyone knows, please do it! After all, there was version 1.1 for 37x37, but it was only cuda version

**Comment by cryptpark** [comment]
10x128 g170e - 10x128 s4588 (KataGo-v1.3, gt 610, 5 sec/move):
7x7  24:26 (g170e won 24 b and 0 w)
9x9  26:24 (10 b and 16 w)
11x11  24:26 (9 b and 15 w)
13x13  25:25 (10 b and 15 w)

**Comment by Friday9i** [comment]
So it is almost exactly iso-strength with the g170 10b net, funny! Thanks

**Comment by cryptpark** [comment]
s5350 may be stronger, I’ll launch in an hour

**Comment by cryptpark** [comment]
200 games 10x128 g170e - 10x128 s5350 (KataGo-v1.3, gt 610, 5 sec/move)
7x7 25:25 (g170e won 25 b and 0 w)
9x9 23:27 (2 b and 21 w)
11x11 28:22 (11 b and 17 w)
13x13 30:20 (11 b and 19 w)

**Comment by cryptpark** [comment]
200 games 10x128 g170e - 10x128 s5350 (KataGo-v1.3, gt 610, 5 sec/move)
9x9 41:59 (14 b and 27 w)
13x13 63:37 (32 b and 31 w)
